### DISCUSSION TITLE
Performance of forwarded Talk ports? Reverse proxy on another machine and Talk HPB?

### DISCUSSION THREAD

**Discussion by neocult-de** [nextcloud/all-in-one#3798] [Questions] [answered]
Hi there,

I have three Nextcloud-AIO instances for testing purposes because the Talk Performance behind reverse proxy is insufficent (nginx proxy manager). Two are each behind their own instance of "nginx proxy manager" (different computer, local network) and one has a dedicated IP and handles everything directly. 

The AIO instances behind npm have their Talk Ports (3478 UDP/TCP) forwareded as Streams. Everything works fine and is reachable in all three instances. I STUN/Turn-Server is available, Talk HPB Signaling is reachable etc., but the performance behind npm as a forwarded Stream has an unbearable delay and therefore low quality. The dedicated IP instance on the same server works smoothly at any time. The other AIO instance behind npm is on different machines and network and have the same slow speed behind npm.

Using another reverse proxy might be an option. I just wanted to have something easily managable, so I don´t have to waste IPv4 addresses for each Service including AIO. Does anyone have experience with that kind of setup? Should I do the forwarding on an iptable basis or another tunnel solution or do I have to have the AIO instance with dedicated IP set? Any hints how to improve performance or how to do this port/stream forwarding from the reverse proxy (different machine) with something like the recommended caddy? The reverse proxy documentation seems to mainly address having it on the same server.

"I looked through dozens and dozens of articles here, but the issue is not being able to connect or configure, but the rather bad performance of Videochat even with two people only. (Ressources on all test machines have been overly generous to eliminate that as the sole issue 16 Cores, 32GB RAM, all on SSD,, Gigabit connections up and down etc.)

Any hints in any direction appreciated.

**Comment by szaimen** [comment, marked as answer]
> The AIO instances behind npm have their Talk Ports (3478 UDP/TCP) forwareded as Streams.

What does that mean? Forwarded as streams using NPM? 

If so, is it working fine if you just open the port 3478 udp and tcp in your router and forward it to that machine and not use NPM in between?

**Comment by neocult-de** [comment]
**>Direct forwarding and masquerading on the reverse-proxy host of the TALK port performs way better and more as originally intended!<**

Forwarding of Ports in Nginx Proxy Manager is the category of "Streams" which is a forward proxy. As Nextcloud-AIO runs behind a reverse-proxy, the forwarding of the TALK port was supposedly working with a forward proxy within the same software. Its performance with latency and bandwith is suboptimal though.



I had the following (please keep in mind that other traffic might get masqueraded and forwarded as well with this ruleset!)
iptables -t nat -A PREROUTING  -p tcp --dport 3478 -j DNAT --to-destination 192.168.0.25
iptables -t nat -A PREROUTING  -p udp --dport 3478 -j DNAT --to-destination 192.168.0.25
iptables -t nat -A POSTROUTING -j MASQUERADE
iptables -A FORWARD -j ACCEPT



Can work behind ipv6 with ip6tables the same way and behind a wireguard connection (then masquerading with "iptables -t nat -A POSTROUTING ! -o %i -j MASQUERADE" in an preup-statement under interfaces in wg-config.

In case anyone else needs a setup like this going. It works of course with other reverse-proxies too, as it is just forwarding the TALK Port only. The reverse proxy need to be on a machine which runs iptables though. (net.ipv4.ip_forward = 1 AND net.ipv6.conf.all.forwarding = 1 in sysctl has to be activated accordingly as well).
